source/extensions/filters/http/composite/action.cc

```diff
@@ -9,6 +9,73 @@ void ExecuteFilterAction::createFilters(Http::FilterChainFactoryCallbacks& callb
   cb_(callbacks);
 }
 
+float ExecuteFilterActionFactory::getSampleRatio(
+    const envoy::extensions::filters::http::composite::v3::ExecuteFilterAction& composite_action) {
+  // In default case, if sample_percent is not populated, sample_ratio is 100%, i.e, all sampled.
+  if (!composite_action.has_sample_percent()) {
+    // 1, i.e, 100%,  means all sampled.
+    return 1;
+  }
+  const auto& sample_percent = composite_action.sample_percent().default_value();
+  float denominator = 100;
+  switch (sample_percent.denominator()) {
+  case envoy::type::v3::FractionalPercent::HUNDRED:
+    denominator = 100;
+    break;
+  case envoy::type::v3::FractionalPercent::TEN_THOUSAND:
+    denominator = 10000;
+    break;
+  case envoy::type::v3::FractionalPercent::MILLION:
+    denominator = 1000000;
+    break;
+  default:
+    throw EnvoyException(
+        fmt::format("ExecuteFilterAction sample_percent config denominator setting "
+                    "is invalid : {}. Valid range 0~2.",
+                    sample_percent.denominator()));
+  }
+
+  float sample_ratio = static_cast<float>(sample_percent.numerator()) / denominator;
+  if (sample_ratio > 1) {
+    throw EnvoyException(fmt::format(
+        "ExecuteFilterAction sample_percent config is invalid. sample_ratio={}(Numerator {} / "
+        "Denominator {}). The valid range is 0~1.",
+        sample_ratio, sample_percent.numerator(), denominator));
+  }
+
+  return sample_ratio;
+}
+
+float ExecuteFilterActionFactory::getSampleRatioRuntime(float sample_ratio,
+                                                        const std::string& runtime_key,
+                                                        Envoy::Runtime::Loader& runtime) {
+  if (!runtime_key.empty()) {
+    // If sample_percent runtime_key is configured, it has to be a number between
+    // [0, 100]. Using 101 to indicate this runtime_key is either not configured or invalid.
+    // Only override the default_value if the runtime_key is in a valid range [0, 100].
+    const uint64_t NO_SAMPLE_PERCENT_RUNTIME = 101;
+    uint64_t sample_percent_runtime =
+        runtime.snapshot().getInteger(runtime_key, NO_SAMPLE_PERCENT_RUNTIME);
+    if (sample_percent_runtime < NO_SAMPLE_PERCENT_RUNTIME) {
+      sample_ratio = static_cast<float>(sample_percent_runtime) / 100;
+    }
+  }
```

Comment:
There are other helpers in runtime that already do this stuff for fractional percent. See https://github.com/envoyproxy/envoy/blob/main/source/common/runtime/runtime_impl.cc.

Reply:
Good idea!  Changed to use existing runtime.snapshot().featureEnabled() API for the sampling check.